modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/AbstractExternalResult.java

```diff
@@ -0,0 +1,392 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.h2.disk;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIO;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIOFactory;
+import org.apache.ignite.internal.processors.query.h2.H2MemoryTracker;
+import org.apache.ignite.internal.processors.query.h2.H2Utils;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.h2.result.ResultExternal;
+import org.h2.result.ResultInterface;
+import org.h2.store.Data;
+import org.h2.value.Value;
+import org.jetbrains.annotations.NotNull;
+
+import static java.nio.file.StandardOpenOption.CREATE_NEW;
+import static java.nio.file.StandardOpenOption.READ;
+import static java.nio.file.StandardOpenOption.WRITE;
+import static org.apache.ignite.internal.processors.query.h2.IgniteH2Indexing.DISK_SPILL_DIR;
+
+/**
+ * Basic class for external result. Contains common methods for file IO.
+ */
+@SuppressWarnings({"MissortedModifiers", "WeakerAccess", "ForLoopReplaceableByForEach"})
+public abstract class AbstractExternalResult implements ResultExternal {
```

Comment:
Could you explain the advantages of this splitting? It is not clear for me. `AbstractExternalResult` encapsulates all file-related operations and it's child classes are responsible for specific serialization and dumping to disk.

Reply:
I tend to agree with Anton here. Concrete external result implementations use _file-related_ operations defined in this class. Implementations does override _file-related_ methods and there is no _template_ methods. Decomposition itself would be a benefit here. If _file-related_ operations were placed in a separate class (e.g. `ExternalResultIO`) it would be obvious that it does not share a state with any `ResultExternal` implementation. Otherwise a code reader should spend a time to figure out mentioned independence (like I did).